crates/fuel-core/src/service/adapters/graphql_api/off_chain.rs

```diff
@@ -187,6 +202,72 @@ impl OffChainDatabase for OffChainIterableKeyValueView {
     fn message_is_spent(&self, nonce: &Nonce) -> StorageResult<bool> {
         self.message_is_spent(nonce)
     }
+
+    fn balance(
+        &self,
+        owner: &Address,
+        asset_id: &AssetId,
+        base_asset_id: &AssetId,
+    ) -> StorageResult<TotalBalanceAmount> {
+        let coins = self
+            .storage_as_ref::<CoinBalances>()
+            .get(&BalancesKey::new(owner, asset_id))?
+            .unwrap_or_default()
+            .into_owned() as TotalBalanceAmount;
+
+        if base_asset_id == asset_id {
+            let messages = self
+                .storage_as_ref::<MessageBalances>()
+                .get(owner)?
+                .unwrap_or_default()
+                .into_owned() as TotalBalanceAmount;
+
+            let total = coins.checked_add(messages).ok_or(anyhow::anyhow!(
+                "Total balance overflow: coins: {coins}, messages: {messages}"
+            ))?;
+
+            debug!(%coins, %messages, total, "total balance");
+            Ok(total)
+        } else {
+            debug!(%coins, "total balance");
+            Ok(coins)
+        }
+    }
+
+    fn balances(
+        &self,
+        owner: &Address,
+        base_asset_id: &AssetId,
+    ) -> StorageResult<BTreeMap<AssetId, TotalBalanceAmount>> {
+        let mut balances = BTreeMap::new();
+        for balance_key in self.iter_all_by_prefix_keys::<CoinBalances, _>(Some(owner)) {
+            let key = balance_key?;
+            let asset_id = key.asset_id();
+
+            let messages = if base_asset_id == asset_id {
```

Comment:
If the user doesn't have any coins with `base_asset_id` but has messages with it, you will not include them in balances because you will not hit this statement.

Reply:
I added test cases for these scenario [here](https://github.com/FuelLabs/fuel-core/pull/2383/files#diff-e83ddeaff43cbea45b7b2a6c5b688054cf32fbcd7c5010fc5059d0812b1f413eR136) and [here](https://github.com/FuelLabs/fuel-core/pull/2383/files#diff-e83ddeaff43cbea45b7b2a6c5b688054cf32fbcd7c5010fc5059d0812b1f413eR178) and fixed the implementation.